### PR TITLE
Fix autoRecover memory leak.

### PR DESCRIPTION
**Pull request by horizonzy**
Descriptions of the changes in this PR:
fixes-#3360

### CODE REVIEW

**Review by hangc0276 — COMMENTED** [review]
We need to close the ledger handle in Auditor#checkAllLedgers() to unregister the listener.
https://github.com/apache/bookkeeper/blob/677ccec3eb84f5be1b3556537871e14eb5e8359c/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java#L1333-L1345

**Comment by horizonzy** [comment]
> We need to close the ledger handle in Auditor#checkAllLedgers() to unregister the listener.
> 
> https://github.com/apache/bookkeeper/blob/677ccec3eb84f5be1b3556537871e14eb5e8359c/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java#L1333-L1345

In ProcessLostFragmentsCb, it will close it.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/AbstractZkLedgerManagerTest.java [hunk 1]
```diff
@@ -824,9 +824,16 @@ public void testUnregisterLedgerMetadataListener() throws Exception {
             ledgerStr, true,
             KeeperException.Code.OK.intValue(), serDe.serialize(metadata), stat);
 
+        mockZkRemoveWatcher();
+
         // unregister the listener
         ledgerManager.unregisterLedgerMetadataListener(ledgerId, listener);
         assertFalse(ledgerManager.listeners.containsKey(ledgerId));
+        assertFalse(watchers.containsKey(ledgerStr));
+        verify(mockZk, times(1)).removeWatches(any(String.class),
```
Comment (zymap) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/AbstractZkLedgerManagerTest.java:833]:
```suggestion
        verify(mockZk, times(1)).removeWatches(eq(getLedgerPath(ledgerId)),
```

Comment (horizonzy) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/AbstractZkLedgerManagerTest.java:833]: good suggestion. 👍

**Review by hangc0276 — APPROVED** [review]
Good job!

**Review by zymap — APPROVED** [review]
LGTM

**Comment by zymap** [comment]
@eolivelli @dlg99 Would you like to help to review this PR? Thanks!

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/AbstractZkLedgerManager.java [hunk 2]
```diff
@@ -154,6 +155,30 @@ private void handleMetadata(Versioned<LedgerMetadata> result, Throwable exceptio
...
+
+        final long ledgerId;
+
+        CancelWatchLedgerMetadataTask(long ledgerId) {
+            this.ledgerId = ledgerId;
+        }
+
+        @Override
+        public void run() {
+            Set<LedgerMetadataListener> listeners = AbstractZkLedgerManager.this.listeners.get(ledgerId);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/AbstractZkLedgerManager.java:171]: this is running in the same thread of "unregisterLedgerMetadataListener"

why do you need a inner class ? can't we simply code this as a regular Java method ?

```
private void cancelLedgerWatchers(long ledgerId) {
.... 
}
```

and call it from unregisterLedgerMetadataListener ?

**Review by eolivelli — CHANGES_REQUESTED** [review]
good catch !

@dlg99 this may explain some problems you have seen in production

**Review by dlg99 — APPROVED** [review]

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/AbstractZkLedgerManagerTest.java (hunk 1 above)
Comment (liuzhuang2017) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/AbstractZkLedgerManagerTest.java]: good idea, 👍

**Comment by zymap** [comment]
@eolivelli Could you please take another look?

**Review by eolivelli — APPROVED** [review]
LGTM

**Comment by eolivelli** [comment]
@hangc0276 please cherry pick to active branches (4.14, 4.15)

**Comment by hangc0276** [comment]
> @hangc0276 please cherry pick to active branches (4.14, 4.15)

@eolivelli  sure.

**Pull request merged**